src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h

```diff
@@ -0,0 +1,156 @@
+      std::vector<ray::BundleSpecification> &bundles,
+      const GcsNodeManager &node_manager) override;
+};
+
+/// GcsPlacementGroupScheduler is responsible for scheduling placement_groups registered
+/// to GcsPlacementGroupManager. This class is not thread-safe.
+class GcsPlacementGroupScheduler : public GcsPlacementGroupSchedulerInterface {
+ public:
+  /// Create a GcsPlacementGroupScheduler
+  ///
+  /// \param io_context The main event loop.
+  /// \param placement_group_info_accessor Used to flush placement_group info to storage.
+  /// \param gcs_node_manager The node manager which is used when scheduling.
+  /// \param schedule_failure_handler Invoked when there are no available nodes to
+  /// schedule placement_groups.
+  /// \param schedule_success_handler Invoked when placement_groups are created on the
+  /// worker successfully.
+  explicit GcsPlacementGroupScheduler(
+      boost::asio::io_context &io_context,
+      std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage,
+      const GcsNodeManager &gcs_node_manager,
+      std::function<void(std::shared_ptr<GcsPlacementGroup>)> schedule_failure_handler,
+      std::function<void(std::shared_ptr<GcsPlacementGroup>)> schedule_success_handler,
+      ReserveResourceClientFactoryFn lease_client_factory = nullptr);
+  virtual ~GcsPlacementGroupScheduler() = default;
+
+  /// Schedule the specified placement_group.
+  /// If there is no available nodes then the `schedule_failed_handler_` will be
+  /// triggered, otherwise the placement_group will be scheduled until succeed or
+  /// canceled.
+  ///
+  /// \param placement_group to be scheduled.
+  void Schedule(std::shared_ptr<GcsPlacementGroup> placement_group) override;
+
+  //   std::deque<std::tuple<BundleSpecification, std::shared_ptr<rpc::GcsNodeInfo>,
+  //                         ReserveResourceCallback>>
+
+  /// Lease resource from the specified node for the specified bundle.
+  void ReserveResourceFromNode();
+
+  /// return resource for the specified node for the specified bundle.
+  ///
+  /// \param bundle A description of the bundle to return.
+  /// \param node The node that the worker will be returned for.
+  void CancelResourceReserve(BundleSpecification bundle_spec,
+                             std::shared_ptr<ray::rpc::GcsNodeInfo> node);
+
+  /// Get an existing lease client or connect a new one.
+  std::shared_ptr<ResourceReserveInterface> GetOrConnectLeaseClient(
+      const rpc::Address &raylet_address);
+
+ protected:
+  /// The io loop that is used to delay execution of tasks (e.g.,
+  /// execute_after).
+  boost::asio::io_context &io_context_;
+  /// Used to update placement group information upon creation, deletion, etc.
+  std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage_;
+  /// Reference of GcsNodeManager.
+  const GcsNodeManager &gcs_node_manager_;
+  /// The handler to handle the scheduling failures.
+  std::function<void(std::shared_ptr<GcsPlacementGroup>)> schedule_failure_handler_;
```

Comment:
So if it can't be placed we immediately return failure and don't retry right?

Reply:
I will add it in the tail of the pending queue and wait for next time to create it.